security-audit/src/main/java/io/micronaut/security/audit/event/listeners/UserAuditingEntityEventListener.java

```diff
@@ -0,0 +1,101 @@
+import io.micronaut.security.authentication.Authentication;
+import io.micronaut.security.utils.SecurityService;
+import jakarta.inject.Singleton;
+
+import java.lang.annotation.Annotation;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Predicate;
+
+/**
+ * An event listener that handles auto-population of entity fields annotated with {@link CreatedBy} or
+ * {@link UpdatedBy} by mapping them from the current {@link Authentication}.
+ *
+ * @author Jeremy Grelle
+ * @since 4.5.0
+ */
+@Singleton
+public class UserAuditingEntityEventListener extends AutoPopulatedEntityEventListener {
+
+    private final SecurityService securityService;
+
+    private final ConversionService conversionService;
+
+    UserAuditingEntityEventListener(SecurityService securityService, ConversionService conversionService) {
+        this.securityService = securityService;
+        this.conversionService = conversionService;
+    }
+
+    @Override
+    public boolean prePersist(@NonNull EntityEventContext<Object> context) {
+        autoPopulateUserIdentity(context, false);
+        return true;
+    }
+
+    @Override
+    public boolean preUpdate(@NonNull EntityEventContext<Object> context) {
+        autoPopulateUserIdentity(context, true);
+        return true;
+    }
+
+    @Override
+    protected @NonNull List<Class<? extends Annotation>> getEventTypes() {
+        return Arrays.asList(PrePersist.class, PreUpdate.class);
+    }
+
+    @Override
+    protected @NonNull Predicate<RuntimePersistentProperty<Object>> getPropertyPredicate() {
+        return property -> {
+            final AnnotationMetadata annotationMetadata = property.getAnnotationMetadata();
+            return annotationMetadata.hasAnnotation(CreatedBy.class) || annotationMetadata.hasAnnotation(UpdatedBy.class);
+        };
+    }
+
+    private void autoPopulateUserIdentity(@NonNull EntityEventContext<Object> context, boolean isUpdate) {
+        final RuntimePersistentProperty<Object>[] applicableProperties = getApplicableProperties(context.getPersistentEntity());
+        for (RuntimePersistentProperty<Object> persistentProperty : applicableProperties) {
+            if (isUpdate) {
+                if (!persistentProperty.getAnnotationMetadata().booleanValue(AutoPopulated.class, AutoPopulated.UPDATEABLE).orElse(true)) {
+                    continue;
+                }
+            }
+
+            final BeanProperty<Object, Object> beanProperty = persistentProperty.getProperty();
+            getCurrentUserIdentityForProperty(beanProperty).ifPresent(identity -> context.setProperty(beanProperty, identity));
+        }
+    }
+
+    private Optional<Object> getCurrentUserIdentityForProperty(BeanProperty<Object, Object> beanProperty) {
+        return securityService.getAuthentication().flatMap(authentication -> conversionService.convert(authentication, beanProperty.getType()));
+    }
```

Comment:
Seems like lots of logic and loops are executed and perhaps do nothing if there is no authentication. Move the check for `securityService.getAuthentication.isPresent()` to an outer most if statement and then loop

Reply:
Good point...done.